Fix `generateTrace` with newer TypeScript versions and `build: false`

mkantor opened the pull request:
Closes #722 and also fixes another issue I noticed along the way: `generateTrace` did not work unless I was using `build: true` (I think that was introduced by #691).

Adds tests and updates documentation.

mkantor committed:
test: add e2e test to validate generateTrace behavior with TS 4.1.x

Also update README.md to better reflect the current state of this
feature.

mkantor committed:
fix: make generateTrace work with newer TypeScript versions

Closes: #722

mkantor committed:
fix: support generateTrace in non-build mode

piotr-oles commented on test/e2e/type-script-tracing.spec.ts:
```diff
@@ -0,0 +1,44 @@
+    { build: false, typescript: '~4.4.0' },
+    { build: true, typescript: '~4.4.0' },
+    { build: false, typescript: '~4.5.0' },
+    { build: true, typescript: '~4.5.0' },
```
I appreciate that you add all these cases, but each case takes time so we have to keep some balance here. Otherwise, our test job will take hours :) I suggest keeping cases only for the min and max versions. I would assume that if it works for these versions, it will work for versions in-between :)

piotr-oles commented on src/typescript/worker/lib/tracing.ts:
```diff
@@ -4,37 +4,48 @@ import { getConfigFilePathFromCompilerOptions } from './config';
-  startTracing(configFilePath: string, traceDirPath: string, isBuildMode: boolean): void;
-  stopTracing(typeCatalog: unknown): void;
-  dumpLegend(): void;
+  startTracing:
```
nit: I think that ‘?’ instead of ‘undefined’ would be easier to read :)

piotr-oles commented on src/typescript/worker/lib/tracing.ts:
```diff
@@ -4,37 +4,48 @@ import { getConfigFilePathFromCompilerOptions } from './config';
+    | undefined
+    | ((tracingMode: Mode, traceDir: string, configFilePath?: string) => void);
+
+  tracing:
```
nit: same here :)

piotr-oles commented on src/typescript/worker/lib/tracing.ts:
```diff
@@ -4,37 +4,48 @@ import { getConfigFilePathFromCompilerOptions } from './config';
 
-// write this type as it's available only starting from TypeScript 4.1.0
+// these types are internal in TypeScript, so reproduce them here
+type Mode = 'project' | 'build' | 'server';
```
nit: could you name it ‘TracingMode’?

piotr-oles reviewed (commented):
Thanks for the PR! I left a few minor comments :)

mkantor committed:
test: eliminate some superfluous test cases for generateTrace

As per code review feedback, do not bother testing intermediate
TypeScript versions.

mkantor commented:
@piotr-oles I addressed your feedback; thanks for the quick review. Let me know if you have any other suggestions.

piotr-oles reviewed (approved):
Thanks, looks good!

piotr-oles commented:
:tada: This PR is included in version 7.2.4 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/fork-ts-checker-webpack-plugin/v/7.2.4)
- [GitHub release](https://github.com/TypeStrong/fork-ts-checker-webpack-plugin/releases/tag/v7.2.4)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
